api.go

```diff
@@ -185,12 +189,108 @@ func (n notAllowedHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 }
 
 type resource struct {
-	resourceType reflect.Type
-	source       CRUD
-	name         string
-	marshalers   map[string]ContentMarshaler
+	resourceType   reflect.Type
+	source         CRUD
+	hooks          CRUDHooks
+	name           string
+	marshalers     map[string]ContentMarshaler
+
+	api            *API
+}
+
+func (res resource) buildRequest(r *http.Request, params httprouter.Params) (*Request, error) {
+	req := Request{
+		PlainRequest: r,
+		Header: r.Header,
+		Params: params,
+		QueryParams: r.URL.Query(),
+		Meta: make(map[string]interface{}),
+		Data: nil,
+	}
+
+	var body []byte
+
+	if r.Body != nil {
+		var err error
+		body, err = ioutil.ReadAll(r.Body)
+		if err != nil {
+			return nil, err
+		}
+		r.Body.Close()
```

Comment:
this should be put 3 lines above and prefixed with a defer like in original `unmarshalRequest`.